src/librustc_mir/interpret/validity.rs

```diff
@@ -580,10 +580,21 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
             }
             ty::Array(tys, ..) | ty::Slice(tys)
                 if {
-                    // This optimization applies only for integer and floating point types
-                    // (i.e., types that can hold arbitrary bytes).
+                    // This optimization applies for types that can hold arbitrary bytes (such as
+                    // integer and floating point types) or for inhabited Zero Sized Types (such as
+                    // the unit type and empty struct types).
                     match tys.kind {
                         ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,
+                        ty::Tuple(t)
+                            if t.len() == 0 || self.ecx.layout_of(tys)?.size == Size::ZERO =>
```

Comment:
The `|| self.ecx.layout_of(tys)?.size == Size::ZERO` part is new, isn't it? And it is wrong: imagine `(!, !)`.

Reply:
Yes, @oli-obk wanted us to try to catch more complex cases that the prior code did not like `struct Foo(());`

Reply:
I think what would work is

```rust
_ => {
    let layout = self.ecx.layout_of(tys)?;
    layout.size == Size::ZERO && !layout.is_uninhabited()
},
```

Reply:
This will also work for any other kind of type, not just structs and tuples, but also

```rust
enum Foo {
   Bar
}
```

or `[T; 0]` or....

Reply:
I am opposed to using `!layout.is_uninhabited()` as indicator for anything. `is_uninhabited` so far can conservatively return `false`, and I don't trust it to serve as a witness for inhabitedness.

Reply:
While I do not think `is_uninhabited` is any kind of conservative, we can also use [is_ty_uninhabited_from_any_module](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.is_ty_uninhabited_from_any_module) which definitely is not conservative.

Reply:
Ok new idea that I think should alleviate your concerns:

Always validate at least one value, even for zsts. Then, if it's a zst, don't validate the other values. Since validating a zst array validates all zsts at the same location and a zst can never have a different value, we'd either error on *all* elements of the array or none. So checking one element for validity checks *all* elements.

Reply:
I have not the slightest idea what `is_ty_uninhabited_from_any_module` does, it talks about forests...

It shouldn't be too hard to write a function that tests `is_inhabited`.

Reply:
Also given that you seem to want to overdesign this, I think "inhabited ZST" is the wrong property. What would still not cover arrays of `(i32, ())`, for example. The right property is "accepts any initialized bit pattern", and we should probably just do a recursive descent over the layout to test for exactly that.

Reply:
Ah, but that's harder than I though due to padding... for `(u32, u16)` we don't want to check if the entire thing is initialized. So maybe that's not the right approach either.